junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala

```diff
@@ -0,0 +1,251 @@
+package org.scalajs.junit.utils
+
+import sbt.testing._
+import org.junit.Test
+
+import scala.annotation.tailrec
+
+abstract class JUnitTest {
+
+  // ANSI colors codes
+  private final val NORMAL = "\u001B[0m"
+  private final val RED = "\u001B[31m"
+  private final val YELLOW = "\u001B[33m"
+  private final val BLUE = "\u001B[34m"
+  private final val MAGENTA = "\u001B[35m"
+  private final val CYAN = "\u001B[36m"
+  private final val GREY = "\u001B[90m"
+
+  private lazy val TIME_TAG = GREY + "<TIME>" + NORMAL
+
+  // appropriate class loader for platform, needs platform extension
+  private val classLoader: ClassLoader =
+    JUnitTestPlatformImpl.getClassLoader
+
+  private val suiteUnderTestName = {
+    val myName = getClass.getName
+    assert(myName.endsWith("Assertions"))
+    myName.stripSuffix("Assertions")
+  }
+
+  protected def frameworkArgss: List[List[String]]
+
+  protected def expectedOutput(context: OutputContext): List[Output]
+
+  val expectedFail = 0
+  val expectedIgnored = 0
+  val expectedTotal = 0
+
+  @Test def testJUnitOutput(): Unit = {
+    for (frameworkArgs <- frameworkArgss) {
+      val jUnitFramework = new com.novocode.junit.JUnitFramework()
+      val runner =
+        jUnitFramework.runner(frameworkArgs.toArray, Array.empty, classLoader)
+      val tasks = runner.tasks(Array(new TaskDef(suiteUnderTestName,
+        jUnitFramework.fingerprints.head, true, Array.empty)))
+
+      val recorder: JUnitTestRecorder =
+        new JUnitTestRecorder(new OutputContext(frameworkArgs))
+
+      // run all tasks and the tasks they generate, needs platform extension
+      JUnitTestPlatformImpl.executeLoop(tasks, recorder)
+
+      recorder.recordDone(runner.done())
+
+      recorder.checkOutput()
+    }
+  }
+
+  protected sealed trait Output
+  private final case class Info(msg: String) extends Output
+  private final case class Warn(msg: String) extends Output
+  private final case class Error(msg: String) extends Output
+  private final case class Debug(msg: String) extends Output
+  private final case class Trace(msg: String) extends Output
+  private final case class Event(status: String) extends Output
+  private final case class Done(msg: String) extends Output
```

Comment:
On `2.12.0-M2` we get the following issue since I added back the `final`, I will have to remove the `final`.
```scala
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:60: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Info(msg: String) extends Output
[error]                            ^
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:61: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Warn(msg: String) extends Output
[error]                            ^
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:62: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Error(msg: String) extends Output
[error]                            ^
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:63: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Debug(msg: String) extends Output
[error]                            ^
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:64: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Trace(msg: String) extends Output
[error]                            ^
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:65: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Event(status: String) extends Output
[error]                            ^
[error] /Users/nastucki/GitHub/scala-js/junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala:66: The outer reference in this type test cannot be checked at run time.
[error]   private final case class Done(msg: String) extends Output
[error]                            ^
```